### DISCUSSION TITLE
njsPC no Data From EasyTouch

### DISCUSSION THREAD

**Discussion by MBinTX** [tagyoureit/nodejs-poolController#870] [App and O/S Config Questions]
Hi Guys,

Just bought a new house a few months ago - my first with a pool.  It's got a Pentair EasyTouch 4, so I've been tinkering with njsPC to get it talking to Home Assistant.  I actually had it working for a while, but now it has stopped, and I've been bashing my head against the wall for days trying to sort it out.  Before I dive in to the problem, here's a quick system rundown:

**System Stuff**
- Home Assistant OS [Core 2023.12.0] running directly on an x86 mini PC.
- njsPC running as an Add-on within HA.
- DashPanel running on a separate Win10 PC. Cloned to Git repository, so it is the latest.  (I can't get DP to run as an HA Add-on)
- njsPC-HA integration being used to land the info into HA.
- Communicating to the EasyTouch via an Elfin EW11 RS485/Ethernet converter.
- Pool controller is a Pentair EasyTouch

![image](https://github.com/tagyoureit/nodejs-poolController/assets/147425098/e7facab7-fd6e-46e0-bbd5-958e49ac8674)

**Config Stuff**
- Default port for njsPC is left at 4200.
- Default port for njsPC-DP is left at 5150
- Elfin EW11 IP is 10.0.0.105, and port is 8899 (its default)
- IP of HA computer is 10.0.0.69.  (DHCP assigned that.  Honestly.)
- IP of computer running DashPanel is 10.0.0.230
- All the IP devices go through a gateway 10.0.0.1.  I omitted it in the sketch above.

**Problem Description:**
Ok, so on to the actual problem.  As mentioned, it was working for a while.  I could see data in HA Dashboards, remotely start/stop the pump etc.   What happens now, is it seems like njsPC isn't interrogating the Pentair; almost like it's waiting.  No data comes in to Home Assistant, and if bypass that via the DashPanel instead, I see what appears to be a sort of default config.  (I think I read a comment somewhere from rstrouse or tagyoureit that you get that default until the first successful communication, then it queries the controller for the actual detail.  Anyway, that's not happening at the moment.)

![image](https://github.com/tagyoureit/nodejs-poolController/assets/147425098/b67c9362-e4aa-4497-9282-1905876a7b7b)

Here's what I've done so far:

1. Verified the connectivity of the EW11 RS485 converter.  It is pingable and healthy on the network.
2. Confirmed the RS485 wiring.  (Nothing changed from when it was working, so likely this is ok).
3. Verified DashPanel is pointed at the njsPC properly.  It's set to 10.0.0.69:4200 per screenshot below.  It does seem to connect to njsPC, though interestingly the auto-discovery via the "Find Server" button does not work.
![image](https://github.com/tagyoureit/nodejs-poolController/assets/147425098/7da95ecc-36bf-4934-a60a-4c2d9b30027c)
4.  Verified that njsPC is pointed at the EW11 correctly.  It's set to 10.0.0.105:8899

![image](https://github.com/tagyoureit/nodejs-poolController/assets/147425098/938d544c-b995-4a2f-81e1-104d68b5dafe)

5.  Checked the njsPC logs in HA.  Seems to be finding the RS485 converter ok?

![image](https://github.com/tagyoureit/nodejs-poolController/assets/147425098/a0f1b553-3de0-47da-8c16-262b78107182)

6. Installed njsPC under npm on my other PC (10.0.0.230) and tried completely bypassing Home Assistant - exact same result.
7. Read every discussion forum I could find.  Helpful but I'm still stuck.  :)

So that's where I am now.  I've rebooted everything multiple times, tried to ensure I have the latest versions of everything, etc.  I'm really scratching my head here.  It feels like I'm just missing one tiny setting or something to kick off the communication, then everything will fall in place.  I just can't figure out what it is!  Would be very appreciative if anyone has insights - thanks!

**Comment by tagyoureit** [comment]
I'm not sure what the error `Error retrieving local ports spawn udevadm ENOENT` means.  There might be something there that it is having trouble connecting to the Elfin or opening the port.

But I would first start off by making sure the Elfin is connected to the EasyTouch correctly.  You can try to connect directly to the Elfin with a TCP serial port monitor and see if you are getting the right data coming through as shown in the [rs485 details](https://github.com/tagyoureit/nodejs-poolController/wiki/RS-485-Adapter-Details) page.

**Comment by MBinTX** [comment]
Thanks for reading my issue and for this suggestion!

I grabbed a copy of Realterm and (I think) managed to make a connection to the Elfin.  All I see for traffic is the "FF" characters repeated.
To be honest, I'm not sure what this is telling me.  I tried a bit of Google to find out what 'FF' represents but couldn't get a clear understanding.  

Could it be indicating that the RS485 channel is open, but there's just no communication happening back and forth?  
Maybe would I need to do something to initiate some traffic on the line?

This is a screenshot of my Realterm.  I added in the config info for the "Display" and "Port" sessions in case I've got it set up wrong.

![image](https://github.com/tagyoureit/nodejs-poolController/assets/147425098/20fe9139-1e80-4796-aac1-c823b64b9117)

**Reply by tagyoureit** [reply]
The EasyTouch is very chatty.  FF is the hex representation of 255.  You should be seeing many more different characters.  Usually, when there are problems we either see nothing or a lot of garbage.  I'll give some thought to as why you are seeing mostly `FF` characters.  Do you have any other USB RS485 that you can test?  I don't personally have the Elfin but I do know a lot of people are successful with it.

**Comment by MBinTX** [comment]
OK, some updates.  Since my last post I:

1. Replaced the wiring and verified all connections.  (Was using CAT5; switched to solid wire).
2. Verified all pinouts correct.
3. Power cycled the Elfin.    
4. Same problem as before.

Then I:

1. Hooked up a wired RS485-USB converter.  (A Waveshare) 
2. Fired up Realterm, and observed a bunch of traffic.  (Screenshot below.)

So now I am starting to suspect the RS-485 side of the Elfin?   Network connectivity to it is fine; it just doesn't want to talk RS485 beyond 'FF'

Now I am trying to bypass Home Assistant completely, and just see if my Win10 PC with njsPC on it (via npm) can talk to the OCP via the wired converter.  If yes, then it all points back to the Elfin.   I've hit a snag though, leading me to my next

**Question:**  What is the syntax I need to use, in the DashPanel, to reference my USB/RS485 converter from my Win10 PC?

The [RS485 Documentation](https://github.com/tagyoureit/nodejs-poolController/wiki/RS-485-Adapter-Details) references a syntax of "dev/ttyUSB0".   That doesn't seem to work.  I've also tried things like "dev/ttyUSB3", "dev/COM3" and just "COM3", but the DashPanel just gives me an error?  (See second screenshot below)

![image](https://github.com/tagyoureit/nodejs-poolController/assets/147425098/de514a7d-e5dc-4994-b8f8-1127e7017261)

![image](https://github.com/tagyoureit/nodejs-poolController/assets/147425098/f819b819-fb9d-4990-b209-6c98a645bb99)

**Comment by MBinTX** [comment]
So I answered my own question on the com port configuration above.   It is simply "COM3" minus the quotes.

For some reason, at first the DP would give me an error when I entered COM3, but after a reboot it took it.

So with the wired RS485/USB converter operational, I fired up a standalone instance of njsPC on my PC, and off it went.  I can see all my data from the OCP as before.

So clearly my problem is with the Elfin.  I'm not sure what went wrong there, as it gives no indication of anything being wrong.  I will either have to get the original Elfin working or get a replacement, as there's no great way to get a wired connection from my OCP to my HA PC, so I need to go wireless.

I'm going to try factory resetting the Elfin and see if that fixes anything.  Out of the box, the only thing I had to do originally to get it working was program an IP address via its software interface (IOTService).   It may start working again, or maybe I got a lemon.  It's only a couple months old.  

Here's a side question, realizing that this is not a Home Assistant forum, but I will ask anyway.  In HA, does anyone know a way to hard-code the settings for the njsPC controller comms, when njsPC is run as an Add-On.  That is, the settings that are normally configured in the DashPanel via _Settings -> Controller -> Comms_.  Right now, if I restart HA, it forgets the settings, so I have to go in via the DashPanel and manually set up the 10.0.0.105 IP address of the Elfin every time.

**Reply by tagyoureit** [reply]
Good luck with the Elfin.  I would try to hardcode any IP addresses.  You mentioned that above and that can always be a troubleshooting headache.  

I would ask the HA question as a new topic.  It likely won't get the attention you want buried in this thread.

**Comment by MBinTX** [comment]
So just to close the loop on this - I am back up and running on my original Elfin.  I'll write the final steps down to help anyone who might read this later with a similar problem.

TLDR; Seemed to be a baud rate issue in the end.

I unhooked the Elfin and then wired it directly to my Waveshare USB RS485 converter.  So I now had two RS485 converters directly wired to each other.  

Then I opened two Realterm sessions on my PC.  For Realterm #1, I connected to the Waveshare, which was connected to my PC via USB (COM3).  For Realterm #2, I connected to the Elfin via an Ethernet connection (i.e. port in Realterm was 10.0.0.105:8899).

I would then type some characters into the one window, and see what I saw in the other.  I did some hex values going back and forth, but they did not seem to jive.  For example, if I pressed the Enter key in Realterm #1, I would see a couple characters show up in Realterm #2.  But if I pressed the Enter key in Realterm #2, I would see _different_ characters show up in Realterm #1.  So it was not symmetrical.

I looked in the settings of the Elfin, and it said the baud rate was set to 115200.  I had set Realterm to 9600, so I was wondering if this was a thing.  I changed the baud rate to 115200 in both Realterm #1 and #2, and after that I started seeing normal values.  For example, if I pressed the 'a' key, I would see a 61 in Hex.  This worked in both directions (#1 --> #2 and #2 --> 1).   

So I went in to the settings of the Elfin and set it down to 9600, reconnected to my OCP, fired up njsPC, and everything worked perfectly!

I'm not sure what to make of the whole baud rate thing.  During the initial install, I don't remember changing the Elfin's baud rate from its  factory default of 115200 (via the IOTService software).  I sort of thought the devices auto-negotiated lower baudrates anyway.  Guess not?   Whatever I did, it was working at the start...

Still not clear how I got to the broken state - if I DID set it down to 9600 during the initial install, I don't know how it got set back up to 115200.  I may have fat-fingered the settings without realizing or something like that.  I guess my lesson learned here it to take better notes during the installation phase.  My second learning is that it always pays to have two RS485 adapters on hand to troubshoot.

Thanks @tagyoureit for your troubleshooting advice!
